File: 8M3W/ehowlqk/boj_1916.js
const fs = require("fs");
const filePath =
  process.platform === "linux" ? "/dev/stdin" : "./8M3W/ehowlqk/boj_1916.txt";
const input = fs.readFileSync(filePath).toString().split("\n");

let idx = 0;

n = Number(input[idx++]);
m = Number(input[idx++]);

const edges = [];
for (idx; idx < m + 2; idx++) {
  edges.push(input[idx].split(" ").map((e) => Number(e)));
}
const [start, end] = input[idx].split(" ").map((e) => Number(e));
console.log(start, end);

const heap = []

File: 8M3W/ehowlqk/boj_1916.py
from heapq import heappop, heappush

n = int(input())
m = int(input())

edges = [list() for _ in range(n + 1)]
for _ in range(m):
    s, e, c = list(map(int, input().split()))
    edges[s].append([c, e])     # 비용, 목적지

start, end = list(map(int, input().split()))
heap = [[0, start]]
distance = [1e10 for _ in range(n + 1)]

while heap:
    dist, curr = heappop(heap)

    if distance[curr] < dist:
        continue
    distance[curr] = dist

    for cost, node in edges[curr]:
        if dist + cost < distance[node]:
            distance[node] = dist + cost
            heappush(heap, [dist + cost, node])

print(distance[end])
